### DISCUSSION TITLE
Apollo Server 3 Subscriptions Connection Context?

### DISCUSSION THREAD

**Discussion by KeithGillette** [apollographql/apollo-server#5468] [Q&A] [answered]
With the removal of integrated Subscription support in Apollo Server 3, how does one add custom properties (such as authentication information) to the operation context argument passed to Subscription Resolver functions? With AS2, one could add custom context via a `subscriptions.onConnect` function on the `ApolloServer` configuration object constructor argument. I don't see custom context covered in [the Reenabling Subscriptions section of the AS3 Migration documentation](https://www.apollographql.com/docs/apollo-server/migration/#reenabling-subscriptions).

**Comment by KeithGillette** [comment, marked as answer]
Answering my own question, I see this is covered in [the Subscriptions section of the Apollo Server documentation](https://www.apollographql.com/docs/apollo-server/data/subscriptions/#operation-context), though not in the Migration section.

**Reply by glasser** [reply]
Actually, I think the section you've linked to wasn't updated for AS3 (did it work for you?). I've filed https://github.com/apollographql/apollo-server/issues/5472 to improve it.

I think that, despite the text in the docs saying "This is not the operation context that’s passed between resolvers" below, the right answer when following our migration guide is to return your context object from `onConnect`.  (AS2 overrode the object returned from onConnect with something that it gets from calling the AS object's context function, but by default I think the object returned from `onConnect` is in fact what subscriptions-transport-ws uses as the execution context object?)

**Reply by KeithGillette** [reply]
Thanks, @glasser. The documentation was "close enough" for me to get it working in our project, as we use TypeScript, which filled in the rest, but you're right that it's not completely up-to-date for AS3: In particular, the `connection` object no longer exists on the `ExpressContext` passed into the `ApolloServer` constructor configuration object, so that needs to be removed from [the ApolloServer configuration example](https://www.apollographql.com/docs/apollo-server/data/subscriptions/#operation-context). Unlike the Subscriptions section of the Migration guide, [The `SubscriptionServer.create` example in the full Subscriptions documentation](https://www.apollographql.com/docs/apollo-server/data/subscriptions/#onconnect-and-ondisconnect) does show `onConnect`, which I guessed and found could be passed same function we used previously to configure Subscription context, which you've already identified in issue #5472.
